### PR TITLE
YAML files don't support inline whitelisting

### PR DESCRIPTION
**Pull request by LouisTrezzini**

### CODE REVIEW

**Comment by domanchi** [comment]
Does this only fail with YAML files? Or does it fail on other file types too?

**Comment by domanchi** [comment]
@LouisTrezzini, please open an issue to report this bug, or complete this pull request to fix it. Pull requests should be used for code wanting to be merged in with the master branch.

File: detect_secrets/plugins/high_entropy_strings.py [hunk 1]
```diff
@@ -151,7 +149,9 @@ def _analyze_yaml_file(self, file, filename):
             # we use this heuristic to quit early if appropriate.
             raise yaml.YAMLError
 
-        data = YamlLineInjector(file).json()
+        parser = YamlFileParser(file)
+        data = parser.json()
+        ignored_lines = parser.get_ignored_lines()
```
[the remark this answers is not in the document]
Comment (LouisTrezzini) [reply detect_secrets/plugins/high_entropy_strings.py]: the current behavior is doing what you suggest:
```
                    if '__line__' in item:
                        potential_secrets.update(
                            self.analyze_string(
                                item['__value__'],
                                item['__line__'],
                                filename,
                            ),
                        )
```

But value is actually the value, not the full line, so the comment is dropped
pyYAML drops comments during preprocessing so we can't use it

I decided to scan the file once to identify all ignored lines and later do a simple O(1) `line in ignored_lines` check

File: detect_secrets/plugins/high_entropy_strings.py [hunk 2]
```diff
@@ -161,13 +161,14 @@ def _analyze_yaml_file(self, file, filename):
...
                 try:
                     if '__line__' in item:
-                        potential_secrets.update(
-                            self.analyze_string(
-                                item['__value__'],
-                                item['__line__'],
-                                filename,
-                            ),
-                        )
+                        if not item['__line__'] in ignored_lines:
```
[the remark this answers is not in the document]
Comment (LouisTrezzini) [reply detect_secrets/plugins/high_entropy_strings.py]: if you look more attentively, the `continue` statement is shared

but I agree it's not very elegant

[the remark this answers is not in the document]
Comment (domanchi) [reply detect_secrets/plugins/high_entropy_strings.py]: Ah, gotcha. ++

File: detect_secrets/plugins/high_entropy_strings.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (domanchi) [reply detect_secrets/plugins/high_entropy_strings.py]: Ahh. Right. Good point.

Can you explain this in the docstring for `get_ignored_lines`? Specifically regarding the fact that the parser drops the comments, and thus, we need to parse the file separately from yaml parsing.

**Comment by LouisTrezzini** [comment]
I pushed the changes you asked, feel free to merge

**Pull request merged**